fix: merge compatible definitions in union types

dmchurch opened the pull request:
Most validation keywords apply to only one of the basic types, so a
"string" type and an "array" type can share a definition without
colliding as a ["string", "array"] type, as long as they don't have any
incompatibilities with each other. Modify UnionTypeFormatter to collapse
these disjoint types into a single definition, without using anyOf.

dmchurch committed:
fix: Merge compatible definitions in union types

Most validation keywords apply to only one of the basic types, so a
"string" type and an "array" type can share a definition without
colliding as a ["string", "array"] type, as long as they don't have any
incompatibilities with each other. Modify UnionTypeFormatter to collapse
these disjoint types into a single definition, without using anyOf.

dmchurch commented:
I've classed this as a "fix" because it doesn't change the semantic meaning of the output, and it doesn't enable new features for schema developers. Feel free to upgrade it to a "feat", though.

domoritz commented on src/Utils/mergeDefinitions.ts:
```diff
@@ -0,0 +1,25 @@
+    if (!type1 || !type2 || uniqueArray(types).length !== types.length) {
+        return null;
+    }
+    const keys = [Object.keys(props1), Object.keys(props2)].flat();
```
```suggestion
    const keys = [...Object.keys(props1), ...Object.keys(props2)];
```

domoritz reviewed (commented):
Thanks for the pull request. Please take a look at the tests.

domoritz commented:
I think you probably just need to update the Vega-Lite test case. You can use the provided script to do it automatically.

dmchurch committed:
Update src/Utils/mergeDefinitions.ts

Change the array.flat() call to use spread notation, since both members are guaranteed to be arrays (unlike the types call, above). Thanks for the suggestion, @domoritz!

Co-authored-by: Dominik Moritz <domoritz@gmail.com>

dmchurch commented:
Right! I'll get on the tests now. Some of them weren't passing even in my original checkout, but at the very least I can make sure that anything affected by this change gets updated.

domoritz commented:
Try `yarn test:update`

dmchurch committed:
Updating test snapshots; fixing reachability tests

The vega-lite test originally worked after updating the snapshot, but
there was a broken reference that got mistakenly purged by
removeUnreachable (which assumed that any definition would
only ever have a single property that could hold references).

Added a validation check to vega-lite.test to register the failure,
updated addReachable to check all possible properties (unless
$ref is set; the spec says that everything else gets ignored in
that case).

dmchurch commented:
I've actually got a couple test cases I want to add to exercise some of the weirder edge cases involving definition collapse, but I'm setting this down for the night. In any event, the existing test cases already exercise it fairly well.

dmchurch committed:
Simplifying logic in addReachable, adding tests

dmchurch committed:
Adding functional tests, mergeDefinitions tests

In addition to the raw schema tests which just check that nothing has
changed since the last snapshot, this adds makeExemplar, which builds
a raw value using the parsed BaseType representation of a TypeScript
type tree. This way, tests can specify valid/invalid data to be verified
using the Ajv validator in addition to raw schema output tests.

dmchurch committed:
Adding enum-merging to mergeDefinitions, unit test

dmchurch committed:
mergeDefinitions works on trivially-mergable types

In a few simple cases like (1 | number), mergeDefinitions will now
correctly collapse two not-quite-disjoint type definitions together,
by discarding the validation restrictions on the more-restricted type.
This replaces the string-merging logic in UnionTypeFormatter, which
was discarding annotations.

dmchurch commented:
Whew! That got _much_ bigger than expected. I added some schema validation tests, which uncovered some issues with the existing code, which required a more elaborate test harness, which means the test harness needs more tests... you get the idea. One of these days I'll get all those yaks shaved.

dmchurch commented:
...and I literally just now realized that the makeExemplar code ought to be living on-or-near the BaseType classes themselves, which would clean up the huge mess of if-else-if-else in makeExemplar.ts. Oh well, there are more revisions I need to make anyway...

Feel free to offer critique/direction if this isn't going in a useful direction, btw, I'm not averse to scrapping things and rebuilding.

dmchurch commented:
...and mergeDefinitions would get so very much smaller if I just called it twice, once in each direction. Yeah, don't merge this yet, I've got work to do.

domoritz commented:
Thank you for working on this and making sure the feature is clean. I definitely think this is a good improvement.

domoritz commented on src/Utils/removeUnreachable.ts:
```diff
@@ -20,39 +30,23 @@ function addReachable(
-        }
-    } else if (definition.not) {
+    } else {
+        addReachable(definition.anyOf, definitions, reachable);
```
This looks great

domoritz reviewed (commented):
Overall, this looks great. Please wrap up the remaining issues and mark the pull request as ready for (final) review.

dmchurch commented on test/utils.ts:
```diff
@@ -8,16 +8,88 @@ import { createFormatter } from "../factory/formatter";
-    return new SchemaGenerator(program, createParser(program, config), createFormatter(config), config);
+interface Validator<T> {
+    (value: T): boolean | Promise<unknown>;
+    //                    ^^^^^^^^^^^^^^^^ *Why* does validateSchema() sometimes return this? Who knows! >.<
```
Haha, that's not us! Ajv has a very unfortunate coding error, the .d.ts file has a different signature from the online documentation. This is just working around that.

I suppose I could open an issue on their repo, though.

domoritz replied:
That would be great and then add a link to it in the code. Thanks!

dmchurch commented on test/vega-lite/schema.json:
```diff
@@ -1558,18 +1554,14 @@
+          "items": {
+            "type": "string"
+          },
+          "type": [
```
This is as intended/expected, actually! The type-specific JSON schema validation constraints specifically only apply when the value is of the appropriate type, which is why a lot of definitions can get collapsed together. If you included "items" without "type" it would mean "The value can be any type, but if it's an array its elements must be strings". So, this schema means "The value can be an array or a string. If the value is an array, its elements must be strings".

That's why all the type-specific properties have different names - "minLength" only applies to strings, but the same concept applied to arrays is "minItems".

domoritz replied:
I see. As long as we can still use the schema to generate APIs from it (which we do), I'm okay.

dmchurch commented on src/Utils/makeExemplar.ts:
```diff
@@ -0,0 +1,230 @@
+import { UnionType } from "../Type/UnionType";
+
+export function makeExemplar(type: BaseType | undefined): unknown {
+    return makeExemplars(type)[0];
```
Yeah, I definitely need to refactor this into something cleaner. On my to-do list, certainly.

domoritz commented:
@dmchurch do you plan to wrap this up soon?

domoritz commented:
@dmchurch can you please finish up this pull request?
